Generator/Sources/NeedleFramework/Generating/Serializers/OutputSerializer.swift

```diff
@@ -37,23 +37,74 @@ class OutputSerializer: Serializer {
     ///
     /// - returns: The source code `String`.
     func serialize() -> String {
-        let registrationBody = providers
-            .map { (provider: SerializedProvider) in
-                provider.registration
-            }
-            .joined()
-            .replacingOccurrences(of: "\n", with: "\n    ")
+        // Dependency factories have deterministic names, so we can only define a
+        // factory once, or the code won't compile.
+        var registeredFactories: Set<String> = []
+        // We generate parent traversal functions to improve compile time on large
+        // codebases. This tells us how many levels deep the tree goes.
+        var maxLevel: Int = 1
 
         let providersSection = providers
             .map { (provider: SerializedProvider) in
-                provider.content
+                if let providerMaxLevel = Int(provider.attributes["maxLevel"] ?? "0") {
+                    if providerMaxLevel > maxLevel {
+                        maxLevel = providerMaxLevel
+                    }
+                }
+                if let factoryName = provider.attributes["factoryName"] {
+                    if registeredFactories.contains(factoryName) {
+                        return ""
+                    }
+                    registeredFactories.insert(factoryName)
+                }
+                return provider.content
             }
             .joined()
+
+        let traversalHelpers = (1...maxLevel).map { num in
+            return """
+            private func parent\(num)(_ component: NeedleFoundation.Scope) -> NeedleFoundation.Scope {
+                return component\(String(repeating: ".parent", count: num))
+            }
+            """
+        }.joined(separator: "\n\n")
         
         let needleDependenciesHash = needleVersionHash.map { return "\"\($0)\""} ?? "nil"
 
         let importsJoined = imports.joined(separator: "\n")
 
+        // With Swift 5.6 and an amd64 target, having a function body that is
+        // thousands of lines long causes a severe compile performance
+        // regression. We were seeing compilation take 4.5x what an x86_64
+        // target would take. As a result, this code splits the calls to
+        // register the dependencies into functions around 50 lines long.
+        // Through some basic testing, this seemed to produce the best results.
+        let linesPerHelper = 50
+        var registrationHelperFuncs: Array<String> = []
```

Comment:
nit: imo `[String]` is more readable: `var registrationHelperFuncs: [String] = []`